src/twisted/test/test_internet.py

```diff
@@ -1363,38 +1365,42 @@ def test_reentrantStreamingProducerUsingWriteSequence(self):
 
 
 
-class PortStringificationTests(unittest.TestCase):
-    if interfaces.IReactorTCP(reactor, None) is not None:
-        def testTCP(self):
-            p = reactor.listenTCP(0, protocol.ServerFactory())
-            portNo = p.getHost().port
-            self.assertNotEqual(str(p).find(str(portNo)), -1,
-                                "%d not found in %s" % (portNo, p))
-            return p.stopListening()
+class PortStringificationTests(TestCase):
+    @skipIf(not interfaces.IReactorTCP(reactor, None),
+            "IReactorTCP is needed")
+    def testTCP(self):
+        p = reactor.listenTCP(0, protocol.ServerFactory())
+        portNo = p.getHost().port
+        self.assertNotEqual(str(p).find(str(portNo)), -1,
+                            "%d not found in %s" % (portNo, p))
+        return p.stopListening()
+
 
-    if interfaces.IReactorUDP(reactor, None) is not None:
-        def testUDP(self):
-            p = reactor.listenUDP(0, protocol.DatagramProtocol())
-            portNo = p.getHost().port
-            self.assertNotEqual(str(p).find(str(portNo)), -1,
-                                "%d not found in %s" % (portNo, p))
-            return p.stopListening()
+    @skipIf(not interfaces.IReactorUDP(reactor, None),
+            "IReactorUDP is needed")
+    def testUDP(self):
+        p = reactor.listenUDP(0, protocol.DatagramProtocol())
+        portNo = p.getHost().port
+        self.assertNotEqual(str(p).find(str(portNo)), -1,
+                            "%d not found in %s" % (portNo, p))
+        return p.stopListening()
 
-    if interfaces.IReactorSSL(reactor, None) is not None and ssl:
-        def testSSL(self, ssl=ssl):
-            pem = util.sibpath(__file__, 'server.pem')
-            p = reactor.listenSSL(0, protocol.ServerFactory(), ssl.DefaultOpenSSLContextFactory(pem, pem))
-            portNo = p.getHost().port
-            self.assertNotEqual(str(p).find(str(portNo)), -1,
-                                "%d not found in %s" % (portNo, p))
-            return p.stopListening()
 
-        if _PY3:
-            testSSL.skip = ("Re-enable once the Python 3 SSL port is done.")
+    @skipIf(not interfaces.IReactorSSL(reactor, None),
+            "IReactorSSL is needed")
+    @skipIf(not ssl, "SSL support is missing")
```

Comment:
I'm not sure, the tests were written like that before.

Reply:
From twisted/python/_setup.py

```
    requirements = [
        "zope.interface >= 4.4.2",
        "constantly >= 15.1",
        "incremental >= 16.10.1",
        "Automat >= 0.3.0",
        "hyperlink >= 17.1.1",
        "PyHamcrest >= 1.9.0",
        "attrs >= 17.4.0",
    ]
```

pyopenssl/cryptography is not a hard requirement.

You need to do the extra `[tls]` install.

```
_EXTRA_OPTIONS = dict(
    dev=_dev,
    tls=[
        'pyopenssl >= 16.0.0',
        # service_identity 18.1.0 added support for validating IP addresses in
        # certificate subjectAltNames
        'service_identity >= 18.1.0',
        # idna 2.3 introduced some changes that break a few things.  Avoid it.
        # The problems were fixed in 2.4.
        'idna >= 0.6, != 2.3',
    ],
```
But I agree that it might be practicle to always have pyopenssl as a dependency.